libsodium-sys/build.rs

```diff
@@ -42,55 +95,406 @@ fn main() {
+    let lib_install_dir = Path::new(&install_dir).join("lib");
+    unwrap!(fs::create_dir_all(&lib_install_dir));
+    let url = format!("{}libsodium-{}-mingw.tar.gz", DOWNLOAD_BASE_URL, VERSION);
+    let compressed_file = download(&url, SHA256);
+
+    // Unpack the tarball
+    let gz_decoder = GzDecoder::new(compressed_file);
+    let mut archive = Archive::new(gz_decoder);
+
+    // Extract just the appropriate version of libsodium.a and headers to the install path
+    let arch_path = if cfg!(target_pointer_width = "32") {
+        Path::new("libsodium-win32")
+    } else if cfg!(target_pointer_width = "64") {
+        Path::new("libsodium-win64")
+    } else {
+        panic!("target_pointer_width not 32 or 64")
+    };
+
+    let unpacked_include = arch_path.join("include");
+    let unpacked_lib = arch_path.join("lib\\libsodium.a");
+    let entries = unwrap!(archive.entries());
+    for entry_result in entries {
+        let mut entry = unwrap!(entry_result);
+        let entry_path = unwrap!(entry.path()).to_path_buf();
+        let full_install_path = if entry_path.starts_with(&unpacked_include) {
+            let include_file = unwrap!(entry_path.strip_prefix(arch_path));
+            Path::new(&install_dir).join(include_file)
+        } else if entry_path == unpacked_lib {
+            lib_install_dir.join("libsodium.a")
+        } else {
+            continue;
+        };
+        unwrap!(entry.unpack(full_install_path));
+    }
+
+    println!(
+        "cargo:rustc-link-search=native={}",
+        lib_install_dir.display()
+    );
+
+    format!("{}/include", install_dir)
+}
+
+#[cfg(not(windows))]
+fn build_libsodium() -> String {
+    use flate2::read::GzDecoder;
+    use std::process::Command;
+    use std::str;
+    use tar::Archive;
+
+    // Determine build target triple
+    let target = unwrap!(env::var("TARGET"));
+
+    // Determine filenames and download URLs
+    let basename = format!("libsodium-{}", VERSION);
+    let url = format!("{}{}.tar.gz", DOWNLOAD_BASE_URL, basename);
+
+    // Determine source and install dir
+    let mut install_dir = get_install_dir();
+    let mut source_dir = unwrap!(env::var("OUT_DIR")) + "/source";
+
+    // Avoid issues with paths containing spaces by falling back to using a tempfile.
+    // See https://github.com/jedisct1/libsodium/issues/207
+    if install_dir.contains(" ") {
+        let fallback_path = "/tmp/".to_string() + &basename + "/" + &target;
+        install_dir = fallback_path.clone() + "/installed";
+        source_dir = fallback_path.clone() + "/source";
+        println!(
+            "cargo:warning=The path to the usual build directory contains spaces and hence \
+             can't be used to build libsodium.  Falling back to use {}.  If running `cargo \
+             clean`, ensure you also delete this fallback directory",
+            fallback_path
+        );
+    }
+
+    // Create directories
+    unwrap!(fs::create_dir_all(&install_dir));
```

Comment:
Do we really need `unwrap` crate?

Reply:
not my code... hesitant to change without understanding implications

Reply:
Read https://crates.io/crates/unwrap . It's the same as .unwrap() but with additional info